src/wallet/walletdb.cpp

```diff
@@ -1059,6 +1059,100 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         }
     }
 
+    // Upgrade to having a wallet hd key
+    // Find the keys which are used in single key internal and external descriptors with
+    // the pre-taproot output types
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_GLOBAL_HD_KEY)
+        ) {
```

Comment:
Some suggestions after our chat:

* let's move this into (a) helper function(s):
  * `FindGlobalHDKeyAmongstDescriptors()`
  * `SetGlobalHDKey()`
* document this function, e.g. explain that it's trying to find the pattern of 3 or 4 descriptor pairs that use the same master key. Note how some might no longer be active because the user imported another descriptor and made it active. There might also be more than one set matching this pattern due to toggling encryption, so we pick the most recent one.
* drop the `tr()` bit; it's not necessary, and it could cause a mismatch if a user did something like manually importing only the receive descriptor
* document the meaning of `WALLET_FLAG_GLOBAL_HD_KEY`. It means we support the feature, but does not guarantee presence of an HD key.
* add (a TODO for) method to the wallet tool to re-run the helper functions; this would be for advanced users who somehow ended up with a wallet not matching this pattern, but have repented, added the right descriptors, and now want the global HD key field set

Reply:
> add (a TODO for) method to the wallet tool to re-run the helper functions; this would be for advanced users who somehow ended up with a wallet not matching this pattern, but have repented, added the right descriptors, and now want the global HD key field set


Is this really needed if we add support for `sethdseed` later?

Reply:
Moved it to a separate function, although I did not split it up into multiple as that does not seem to be actually useful. Also added some additional comments.